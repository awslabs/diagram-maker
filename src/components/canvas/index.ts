export { default as Canvas, CanvasProps } from './Canvas';
