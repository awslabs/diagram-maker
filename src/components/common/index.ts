export { default as ComposeView } from './ComposeView';
