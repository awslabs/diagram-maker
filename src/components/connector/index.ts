export {
  default as Connector, ConnectorProps, ConnectorType
} from './Connector';
