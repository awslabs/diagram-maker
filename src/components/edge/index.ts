export { EdgeStyle } from './EdgeCurve';
export { default as Edge, EdgeProps } from './Edge';
export { default as EdgeBadge, EdgeBadgeProps } from './EdgeBadge';
export { default as PotentialEdge, PotentialEdgeProps } from './PotentialEdge';
