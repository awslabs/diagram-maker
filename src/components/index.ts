export { Canvas, CanvasProps } from './Canvas';
export { default as ComposeView } from './common/ComposeView';
export {
  default as Connector,
  ConnectorProps,
  ConnectorType
} from './connector/Connector';
export {
  default as ContextMenu,
  ContextMenuProps
} from './contextMenu/ContextMenu';
export { Node, NodeProps, PotentialNode, PotentialNodeProps } from './node';
export { Panel, PanelProps } from './panel';
export { SelectionMarquee, SelectionMarqueeProps } from './selectionMarquee';
export { View, ViewProps } from './view/View';
export { getConnectedView, ConnectedViewOwnProps } from './view/ConnectedView';
export { Workspace, WorkspaceProps } from './Workspace';

export * from './edge/EdgeCurve';
export { default as Edge, EdgeProps } from './edge/Edge';
export { default as EdgeBadge, EdgeBadgeProps } from './edge/EdgeBadge';
export {
  default as PotentialEdge,
  PotentialEdgeProps
} from './edge/PotentialEdge';

export {
  default as GridPattern,
  GridPatternProps
} from './canvas/patterns/GridPattern';

export {
  default as DotPattern,
  DotPatternProps
} from './canvas/patterns/DotPattern';

export { render, destroy } from './renderUtils';
