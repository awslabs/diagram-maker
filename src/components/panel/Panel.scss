@import '~diagramMaker/styles/mixins';

.dm-panel {
  @include dm-container-shadow;
  position: absolute;
  background-color: white;
  overflow: hidden;
  .dm-content {
    padding: 20px;
  }
}
