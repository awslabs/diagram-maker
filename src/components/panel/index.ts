export { default as Panel, PanelProps } from './Panel';
