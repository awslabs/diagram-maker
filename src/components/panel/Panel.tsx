import * as classnames from 'classnames';
import * as Preact from 'preact';

import { ComposeView } from 'diagramMaker/components/common';
import { BoundRenderCallback, DestroyCallback } from 'diagramMaker/service/ConfigService';
import { DiagramMakerComponentsType } from 'diagramMaker/service/ui/types';
import { Position, PositionAnchorType, Size  } from 'diagramMaker/state/types';

import './Panel.scss';

export interface PanelProps {
  id: string;
  size: Size;
  position: Position;
  renderCallback: BoundRenderCallback;
  destroyCallback: DestroyCallback;
  viewContainerSize: Size;
  positionAnchor?: PositionAnchorType;
}

const getPosition = (
  x: number,
  y: number,
  positionAnchor: PositionAnchorType,
  panelSize: Size,
  viewContainerSize: Size
): Position => {
  switch (positionAnchor) {
    case PositionAnchorType.TOP_LEFT:
      return { x, y };
    case PositionAnchorType.TOP_RIGHT:
      const topRightX = viewContainerSize.width - panelSize.width - x;
      const topRightY = y;

      return { x: topRightX, y: topRightY };
    case PositionAnchorType.BOTTOM_LEFT:
      const bottomLeftX = x;
      const bottomLeftY = viewContainerSize.height - panelSize.height - y;

      return { x: bottomLeftX, y: bottomLeftY };
    case PositionAnchorType.BOTTOM_RIGHT:
      const bottomRightX = viewContainerSize.width - panelSize.width - x;
      const bottomRightY = viewContainerSize.height - panelSize.height - y;

      return { x: bottomRightX, y: bottomRightY };
  }
};

export default class Panel extends Preact.Component<PanelProps> {

  public render(): JSX.Element {
    const { x, y } = this.props.position;
    const { width, height } = this.props.size;
    const positionAnchor = this.props.positionAnchor || PositionAnchorType.TOP_LEFT;
    const { destroyCallback, renderCallback, viewContainerSize } = this.props;

    const { x: translateX, y: translateY } = getPosition(x, y, positionAnchor, this.props.size, viewContainerSize);
    const transform = `translate3d(${translateX}px, ${translateY}px, 0)`;

    return (
      <div
        data-id={this.props.id}
        data-type={DiagramMakerComponentsType.PANEL}
        data-event-target={true}
        className="dm-panel"
        style={{ width, height, transform }}
      >
        <ComposeView
          renderCallback={renderCallback}
          destroyCallback={destroyCallback}
        />
      </div>
    );
  }
}
