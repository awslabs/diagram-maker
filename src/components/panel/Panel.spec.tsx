import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import { Panel } from 'diagramMaker/components/panel';
import { PositionAnchorType } from 'diagramMaker/state/types';

describe('Panel', () => {
  const destroyCallback = jest.fn();
  const renderCallback = jest.fn((container: HTMLElement): HTMLElement => {
    const element = document.createElement('div');
    element.textContent = 'Example panel content';
    container.appendChild(element);
    return element;
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders at the given position and size', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        position={{ x: 400, y: 600 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at top left with correct size', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.TOP_LEFT}
        position={{ x: 0, y: 0 }}
        size={{ width: 300, height: 600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at bottom left with correct size', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.BOTTOM_LEFT}
        position={{ x: 0, y: 0 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at top right with correct size', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.TOP_RIGHT}
        position={{ x: 0, y: 0 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at bottom right with correct size', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.BOTTOM_RIGHT}
        position={{ x: 0, y: 0 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at top left with correct size and relative position', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.TOP_LEFT}
        position={{ x: 100, y: 100 }}
        size={{ width: 300, height: 600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at bottom left with correct size and relative position', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.BOTTOM_LEFT}
        position={{ x: 100, y: 100 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at top right with correct size and relative position', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.TOP_RIGHT}
        position={{ x: 100, y: 100 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });

  it('renders panel docked at bottom right with correct size and relative position', () => {
    const panel = shallow(
      <Panel
        id="myPanel"
        positionAnchor={PositionAnchorType.BOTTOM_RIGHT}
        position={{ x: 100, y: 100 }}
        size={{ width:300, height:600 }}
        viewContainerSize={{ width: 1200, height: 800 }}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(panel).toMatchSnapshot();
  });
});
