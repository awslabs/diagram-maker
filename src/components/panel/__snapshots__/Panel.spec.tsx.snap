// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders at the given position and size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(400px, 600px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at bottom left with correct size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(0px, 200px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at bottom left with correct size and relative position 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(100px, 100px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at bottom right with correct size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(900px, 200px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at bottom right with correct size and relative position 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(800px, 100px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at top left with correct size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(0px, 0px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at top left with correct size and relative position 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(100px, 100px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at top right with correct size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(900px, 0px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Panel renders panel docked at top right with correct size and relative position 1`] = `
preact-render-spy (1 nodes)
-------
<div
  data-id="myPanel"
  data-type="DiagramMaker.Panel"
  data-event-target={true}
  class="dm-panel"
  style="width: 300px; height: 600px; transform: translate3d(800px, 100px, 0);"
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;
